File: android/deeplinkSDK/src/main/kotlin/com/deeplink/sdk/DeeplinkData.kt
package com.deeplink.sdk

/**
 * Data returned by the SDK when a deep link is resolved.
 */
data class DeeplinkData(
    val linkId: String,
    val alias: String,
    val iosUrl: String?,
    val androidUrl: String?,
    val destinationUrl: String,
    val utmSource: String?,
    val utmMedium: String?,
    val utmCampaign: String?,
    val utmContent: String?,
    val utmTerm: String?,
    /** Custom key-value metadata set on the link in the dashboard. */
    val metadata: Map<String, String> = emptyMap(),
    /** Creative name for attribution reporting. */
    val creativeName: String? = null,
    /** Creative ID for attribution reporting. */
    val creativeId: String? = null,
)

/**
 * Result returned after successfully creating a deep link via the SDK.
 * @property url Full short URL (e.g. `https://dl.yourapp.com/abc123`).
 * @property alias Short alias component (e.g. `abc123`).
 * @property linkId Server-assigned link UUID.
 */
data class CreatedLink(
    val url: String,
    val alias: String,
    val linkId: String,
)

/** Parsed incoming Android App Link or custom scheme URL. */
data class IncomingLink(
    val url: android.net.Uri,
    val pathSegments: List<String>,
    val params: Map<String, String>,
)

File: ios/Sources/DeeplinkSDK/DeeplinkData.swift
import Foundation

/// Data returned by the SDK when a deep link is resolved.
public struct DeeplinkData {
    public let linkId: String
    public let alias: String
    public let iosUrl: String?
    public let androidUrl: String?
    public let destinationUrl: String
    public let utmSource: String?
    public let utmMedium: String?
    public let utmCampaign: String?
    public let utmContent: String?
    public let utmTerm: String?
    /// Custom key-value metadata set on the link in the dashboard.
    public let metadata: [String: String]
    /// Creative name for attribution reporting.
    public let creativeName: String?
    /// Creative ID for attribution reporting.
    public let creativeId: String?
}

/// Result returned after successfully creating a deep link via the SDK.
public struct CreatedLink {
    /// The full short URL (e.g. `https://dl.yourapp.com/abc123`).
    public let url: String
    /// The short alias component (e.g. `abc123`).
    public let alias: String
    /// The server-assigned link UUID.
    public let linkId: String
}

// MARK: - Internal Codable representation

struct SDKCreateLinkResponse: Decodable {
    let url: String
    let alias: String
    let linkId: String
}

struct SDKInitResponse: Decodable {
    let matched: Bool
    let data: SDKInitData?
}

struct SDKInitData: Decodable {
    let linkId: String
    let alias: String
    let iosUrl: String?
    let androidUrl: String?
    let destinationUrl: String
    let utmSource: String?
    let utmMedium: String?
    let utmCampaign: String?
    let utmContent: String?
    let utmTerm: String?
    let metadata: [String: String]?
    let creativeName: String?
    let creativeId: String?

    func toDeeplinkData() -> DeeplinkData {
        DeeplinkData(
            linkId: linkId,
            alias: alias,
            iosUrl: iosUrl,
            androidUrl: androidUrl,
            destinationUrl: destinationUrl,
            utmSource: utmSource,
            utmMedium: utmMedium,
            utmCampaign: utmCampaign,
            utmContent: utmContent,
            utmTerm: utmTerm,
            metadata: metadata ?? [:],
            creativeName: creativeName,
            creativeId: creativeId
        )
    }
}
